src/core/Rule.ml

```diff
@@ -287,6 +313,52 @@ let get_sink_requires { sink_requires; _ } =
   | None -> PLabel default_source_label
   | Some { precondition; _ } -> precondition
 
+(* Check if a formula has the following shape:
+ *
+ *     patterns:
+ *     - pattern-either:
+ *       - patterns:
+ *         - pattern-inside: |
+ *             <P>
+ *         - pattern: <func>(<args>)
+ *       - ...
+ *     - focus-metavariable: $MVAR
+ *
+ * See 'taint_sink', field 'sink_is_func_with_focus'. *)
+let is_func_sink_with_focus sink_formula =
```

Comment:
I've thought a bit more. I think the reason we need `is_inside_or_not` is so that we can filter them out. We filter them out because we don't want to short-circuit the `for_all`.

One possibility to improve readability may be to have a {`true`, `dont_care`} type.

`true && dont_care` is equivalent to `true || dont_care` is equivalent to `true`

`Not` and `Inside` are `dont_care`s.

`Call`'s are `true`.

`Or`s and `And`s are folds with the initial value being `dont_care`, and if at the end the folded value is true, then it's a more precise sink.

Not sure if this works, and not sure if it makes things more readable. Just thinking out loud.

Reply:
I actually think it's more clear having a separate `is_inside_or_not`. In a `patterns:` I am not invested in `pattern-inside:`, `pattern-not-inside:` and `pattern-not:` because they just provide context to the match, I am interesting in checking what it is actually going to give me the matches, to check that is a func-call pattern. Tried to clarify this in the comments.